akka-cluster-sharding/src/test/scala/akka/cluster/sharding/LeastShardAllocationStrategy2Spec.scala

```diff
@@ -0,0 +1,114 @@
+  private val regionC = system.actorOf(Props.empty, "regionC")
+
+  private val shards = (1 to 999).map(n => ("00" + n.toString).takeRight(3))
+
+  def createAllocations(aCount: Int, bCount: Int = 0, cCount: Int = 0): Map[ActorRef, Vector[String]] = {
+    Map(
+      regionA -> shards.take(aCount).toVector,
+      regionB -> shards.slice(aCount, aCount + bCount).toVector,
+      regionC -> shards.slice(aCount + bCount, aCount + bCount + cCount).toVector)
+  }
+
+  private val strategyWithoutLimits = new LeastShardAllocationStrategy2(absoluteLimit = 1000, relativeLimit = 1.0)
+
+  "LeastShardAllocationStrategy2" must {
+    "allocate to region with least number of shards" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 1, bCount = 1)
+      allocationStrategy.allocateShard(regionA, "003", allocations).futureValue should ===(regionC)
+    }
+
+    "rebalance shards [1, 2, 0]" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 1, bCount = 2)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set("002"))
+    }
+
+    "rebalance shards [2, 0, 0]" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 2)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set("001"))
+    }
+
+    "not rebalance shards [1, 1, 0]" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 1, bCount = 1)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set.empty[String])
+    }
+
+    "rebalance shards [3, 0, 0]" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 3)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set("001", "002"))
+    }
+
+    "rebalance shards [4, 4, 0]" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 4, bCount = 4)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set("001", "005"))
+    }
+
+    "rebalance shards [4, 4, 2]" in {
+      // this is handled by phase 2, to find diff of 2
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 4, bCount = 4, cCount = 2)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set("001"))
+    }
+
+    "rebalance shards [5, 5, 0]" in {
+      val allocationStrategy = strategyWithoutLimits
+      val allocations = createAllocations(aCount = 5, bCount = 5)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set("001", "006"))
+
+      // so far [4, 4, 2]
+      val allocations2 = createAllocations(aCount = 4, bCount = 4, cCount = 2)
```

Comment:
It would be nice to have a small test utility to take the original `allocations` and rebalance it with the results of the first pass. It will make it easier to follow the test. 

Then we could assert: 

On first pass, we should move "001" and "006" and end up with `[4, 4, 2]`. 

Second pass should move "002" and end up with `[3, 4, 3]`.

Reply:
Good idea. I already have that in the randomized test. Will include here also.

Reply:
done: 221f03a